prow.sh

```diff
@@ -0,0 +1,667 @@
+#! /bin/bash -e
+#
+# Copyright 2019 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# This script runs inside a Prow job. It runs unit tests ("make test")
+# and E2E testing. This E2E testing covers different scenarios (see
+# https://github.com/kubernetes/enhancements/pull/807):
+# - running the stable hostpath example against a Kubernetes release
+# - running the canary hostpath example against a Kubernetes release
+# - building the component in the current repo and running the
+#   stable hostpath example with that one component replaced against
+#   a Kubernetes release
+#
+# The intended usage of this script is that individual repos import
+# csi-release-tools, then link their top-level prow.sh to this or
+# include it in that file. When including it, several of the variables
+# can be overridden in the top-level prow.sh to customize the script
+# for the repo.
+#
+# The expected environment is:
+# - $GOPATH/src/<import path> for the repository that is to be tested,
+#   with PR branch merged (when testing a PR)
+# - running on linux-amd64
+# - bazel installed (when testing against Kubernetes master), must be recent
+#   enough for Kubernetes master
+# - kind (https://github.com/kubernetes-sigs/kind) installed
+# - optional: Go already installed
+
+# Sets the default value for a variable if not set already and logs the value.
+# Any variable set this way is usually something that a repo's .prow.sh
+# or the job can set.
+configvar () {
+    # Ignore: Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"?
+    # shellcheck disable=SC2140
+    eval : \$\{"$1":="\$2"\}
+    eval echo "\$3:" "$1=\${$1}"
+}
+
+# Go version to use. If the pre-installed Go is missing or a different
+# version, this version here will get installed.
+configvar CSI_PROW_GO_VERSION 1.11.4 "Go version"
```

Comment:
Primarily it is for building the component itself. But it also gets used to build kind (only if there is no pre-built binary available, so this should become the exception), the E2E test suite, and Kubernetes (when testing against a floating  version like master or "latest release-1.13").

Hmm, this can indeed become problematic. My main concern were "go fmt" checks in the component itself, because those are known to be dependent on very specific Go versions. But Kubernetes building will soon fail for Go < 1.12. I'll change that so that each usage of Go can potentially have its own version.

Reply:
Can we not use the go version from the common makefile to build the component and check go fmt?

Is there a way we can dynamically find the go version needed to build each component so we don't have to hardcode it in our test scripts?